### PR TITLE
feat(redis): Allow configuring Redis pools individually

### PR DESCRIPTION
**Pull request by loewenheim**
Implements #3829.

### CODE REVIEW

File: relay-config/src/redis.rs [hunk 1]
```diff
@@ -86,7 +87,8 @@ pub enum RedisConnection {
 }
 
 /// Configuration for connecting a redis client.
-#[derive(Clone, Debug, Serialize, Eq, PartialEq)]
+#[derive(Clone, Debug, Deserialize, Serialize, Eq, PartialEq)]
+#[serde(from = "RedisConfigFromFile")]
```
Comment (loewenheim) [inline relay-config/src/redis.rs:91]: This is equivalent to the manual `Deserialize` impl from before.

File: relay-config/src/config.rs [hunk 2]
```diff
@@ -2273,25 +2273,15 @@ impl Config {
...
-                .max_connections
-                .unwrap_or(cpu_concurrency as u32 * 2)
-                .min(crate::redis::DEFAULT_MIN_MAX_CONNECTIONS),
-            connection_timeout: redis.options.connection_timeout,
-            max_lifetime: redis.options.max_lifetime,
-            idle_timeout: redis.options.idle_timeout,
-            read_timeout: redis.options.read_timeout,
-            write_timeout: redis.options.write_timeout,
-        };
+    pub fn redis(&self) -> Result<RedisPools, RedisError> {
```
Comment (loewenheim) [inline relay-config/src/config.rs:2276]: Returning `RedisPools` here introduces some coupling between `relay_config` and `relay_redis`. I previously had an auxiliary type here that kept everything local, but I think it's not worth the additional indirection and ceremony.

File: relay-config/src/redis.rs [hunk 3]
```diff
@@ -128,12 +130,98 @@ impl From<RedisConfigFromFile> for RedisConfig {
...
+) -> Result<relay_redis::RedisPools, RedisError> {
+    match configs {
+        RedisConfigs::Unified(cfg) => {
+            let pool = create_redis_pool(cfg, cpu_concurrency)?;
+            Ok(RedisPools {
+                project_configs: Some(pool.clone()),
+                cardinality: Some(pool.clone()),
+                quotas: Some(pool.clone()),
+                misc: Some(pool),
+            })
```
Comment (loewenheim) [inline relay-config/src/redis.rs:193]: We could also create each pool individually here, and will have to when we want to differentiate between sync and async pools.

File: relay-server/src/service.rs [hunk 4]
```diff
@@ -343,6 +340,32 @@ impl ServiceState {
     }
 }
 
+fn create_redis_pool(
+    connection: &RedisConnection,
+    options: RedisConfigOptions,
+) -> Result<RedisPool, RedisError> {
```
Comment (Dav1dde) [inline relay-server/src/service.rs]: If this takes a tuple, passing in values should be a bit nicer:
```suggestion
fn create_redis_pool(
    (connection, options): (&RedisConnection, RedisConfigOptions)
) -> Result<RedisPool, RedisError> {
```

File: relay-config/src/redis.rs [hunk 5]
```diff
@@ -128,12 +130,90 @@ impl From<RedisConfigFromFile> for RedisConfig {
...
+        RedisConfigs::Individual {
+            project_configs,
+            cardinality,
+            quotas,
+            misc,
+        } => {
+            let project_configs = create_redis_pool(project_configs, cpu_concurrency);
+            let cardinality = create_redis_pool(cardinality, cpu_concurrency);
+            let quotas = create_redis_pool(quotas, cpu_concurrency);
+            let misc = create_redis_pool(misc, cpu_concurrency);
```
Comment (Dav1dde) [inline relay-config/src/redis.rs]: Only the project config pool lives outside the processor with a pretty much unbounded concurrency, so for compatibility sake we should keep the old limit here. All the other pools are limited by the concurrency defined, so we can just use that as an upper bound.

We can also only use the `min()` from before for the project config pool for the same reason.

File: relay-config/src/redis.rs [hunk 6]
```diff
@@ -128,12 +130,90 @@ impl From<RedisConfigFromFile> for RedisConfig {
...
+pub(super) fn create_redis_pool(
+    config: &RedisConfig,
+    cpu_concurrency: usize,
+) -> (&RedisConnection, RedisConfigOptions) {
+    let options = RedisConfigOptions {
+        max_connections: config
+            .options
+            .max_connections
+            .unwrap_or(cpu_concurrency as u32 * 2)
+            .min(crate::redis::DEFAULT_MIN_MAX_CONNECTIONS),
```
Comment (Dav1dde) [inline relay-config/src/redis.rs]:
```suggestion
pub(super) fn create_redis_pool(
    config: &RedisConfig,
    default_connections: usize,
) -> (&RedisConnection, RedisConfigOptions) {
    let options = RedisConfigOptions {
        max_connections: config
            .options
            .max_connections
            .unwrap_or(default_connections)
```

If the function takes `default_connections` and we move the logic out, we can have more easily different defaults for per pool.

File: relay-server/src/services/stats.rs [hunk 7]
```diff
@@ -100,18 +100,36 @@ impl RelayStats {
         }
     }
 
+    #[cfg(feature = "processing")]
+    async fn redis_pool(redis_pool: &RedisPool, name: &str) {
```
Comment (Dav1dde) [inline relay-server/src/services/stats.rs]: This doesn't need to be async. The other 'handler' is async just so the `join` works without being too annoying.

File: relay-server/src/services/stats.rs [hunk 8]
```diff
@@ -100,18 +100,36 @@ impl RelayStats {
...
-        metric!(gauge(RelayGauges::RedisPoolConnections) = u64::from(state.connections));
-        metric!(gauge(RelayGauges::RedisPoolIdleConnections) = u64::from(state.idle_connections));
+    async fn redis_pools(&self) {
+        if let Some(RedisPools {
+            project_configs,
+            cardinality,
+            quotas,
+            misc,
+        }) = self.redis_pools.as_ref()
+        {
```
Comment (Dav1dde) [inline relay-server/src/services/stats.rs:127]: Why the additional scope?

File: relay-server/src/services/processor.rs [hunk 9]
```diff
@@ -1151,32 +1151,41 @@ impl EnvelopeProcessorService {
...
+        };
+
         let inner = InnerProcessor {
             pool,
             global_config,
             cogs,
             #[cfg(feature = "processing")]
-            redis_pool: redis.clone(),
+            // TODO: Tentatively using `quotas` for this
+            redis_pool: quotas.clone(),
```
Comment (Dav1dde) [inline relay-server/src/services/processor.rs]: Can't we already rename this field to `quotas_pool`?

File: relay-config/src/redis.rs (hunk 1 above)
Comment (Dav1dde) [reply relay-config/src/redis.rs:91]: Nice

File: relay-server/src/services/stats.rs (hunk 8 above)
Comment (loewenheim) [reply relay-server/src/services/stats.rs:127]: This is the body of an `if-let`. Unless I'm misunderstanding the question?

File: relay-config/src/redis.rs (hunk 5 above)
Comment (loewenheim) [reply relay-config/src/redis.rs]: Sorry, can you rephrase that? What should the `default_connections` be for each of these pools? `min(2 * cpu_concurrency, DEFAULT_MIN_MAX_CONNECTIONS)` for the `project_configs` pool, just `2 * cpu_concurrency` for the rest? Or am I misunderstanding?

File: relay-server/src/services/stats.rs (hunk 8 above)
Comment (Dav1dde) [reply relay-server/src/services/stats.rs:127]: Ah sorry, github rendered this very weird.

File: relay-config/src/redis.rs (hunk 5 above)
Comment (Dav1dde) [reply relay-config/src/redis.rs]: Assuming `create_redis_pools` takes the default number of connections (instead of concurrency):

```rs
            let project_configs = create_redis_pool(
                project_configs, 
                (cpu_concurrency * 2).min(crate::redis::DEFAULT_MIN_MAX_CONNECTIONS)
            );
            let cardinality = create_redis_pool(cardinality, cpu_concurrency);
            let quotas = create_redis_pool(quotas, cpu_concurrency);
            let misc = create_redis_pool(misc, cpu_concurrency);
```

And for the unified part:

```rs
        RedisConfigs::Unified(cfg) => {
            let default_connections = (cpu_concurrency * 2).min(crate::redis::DEFAULT_MIN_MAX_CONNECTIONS);
            let pool = create_redis_pool(cfg, default_connections);
```

Comment (loewenheim) [reply relay-config/src/redis.rs]: It should be noted that the original code was
```rust
max_connections: redis
        .options
        .max_connections
        .unwrap_or(cpu_concurrency as u32 * 2)
        .min(crate::redis::DEFAULT_MIN_MAX_CONNECTIONS),
```
which does something different (it always clamps the connections to `DEFAULT_MIN_MAX_CONNECTIONS`, not just when `max_connections` isn't set).

**Review by Dav1dde — APPROVED** [review]
